tensorflow/python/data/ops/dataset_ops.py

```diff
@@ -3675,19 +3676,65 @@ class Options(options_lib.OptionsBase):
       "`tf.data.ThreadingOptions` for more details.",
       default_factory=threading_options.ThreadingOptions)
 
+  autotune = options_lib.create_option(
+      name="autotune",
+      ty=autotune_options.AutotuneOptions,
+      docstring="The autotune options associated with the dataset. See "
+      "`tf.data.AutotuneOptions` for more details.",
+      default_factory=autotune_options.AutotuneOptions)
+
+  def _get_optimization_autotune_opt_map(self):
+    """Helper method which returns the backward-compatibility
+    attribute mapping between `experimental_optimization` and
+    `autotune` options.
+    """
+    return {
+        # experimental_optimization : autotune
+        "autotune": "enabled",
+        "autotune_buffers": "experimental_autotune_buffers",
+        "autotune_cpu_budget": "cpu_budget",
+        "autotune_ram_budget": "ram_budget"
+    }
+
   def __getattr__(self, name):
+    # handle backward compatibility with deprecated options
     if name == "experimental_threading":
       logging.warning("options.experimental_threading is deprecated. "
                       "Use options.threading instead.")
       return getattr(self, "threading")
+    elif name == "experimental_optimization":
+      value = getattr(self, "experimental_optimization")
+      autotune_opt = getattr(self, "autotune")
+      attr_map = self._get_optimization_autotune_opt_map()
+      for old_attr, new_attr in attr_map.items():
+        logging.warning("options.experimental_optimization.{}"
+        " is deprecated. Use options.autotune.{} instead".format(
+          old_attr, new_attr))
+        res = getattr(autotune_opt, new_attr)
+        if res:
+          setattr(value, old_attr, res)
+      return value
```

Comment:
For similar reasons that prompted my changes to ThreadingOptions, we do not want to keep both the old and the new attributes and try to keep them in sync. Instead, we should remove the old attributes and rerouted getting and setting of the old attributes to the new attributes. 

You can accomplish this by having AutotuneOptions be a private attribute of the OptimizationOptions and have all accesses of `autotune` be routed through `OptimizationOptions` which either returns its `AutotuneOptions` instance (if user is accessing `autotune`) or updates the `AutotuneOptions` instance, if user is trying to update any of the removed attributes.

Reply:
addressed in the commit.